core/shared/src/main/scala/zio/stm/TReentrantLock.scala

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright 2017-2019 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio.stm
+
+import zio.{ Fiber, Managed }
+
+import TReentrantLock._
+
+/**
+ * A `TReentrantLock` is a reentrant read/write lock. Multiple readers may all
+ * concurrently acquire read locks. Only one writer is allowed to acquire a
+ * write lock at any given time. Read locks may be upgraded into write locks.
+ * A fiber that has a write lock may acquire other write locks or read locks.
+ *
+ * The two primary methods of this structure are `readLock`, which acquires a
+ * read lock in a managed context, and `writeLock`, which acquires a write lock
+ * in a managed context.
+ *
+ * Although located in the STM package, there is no need for locks within
+ * STM transactions. However, this lock can be quite useful in effectful code,
+ * to provide consistent read/write access to mutable state; and being in STM
+ * allows this structure to be composed into more complicated concurrent
+ * structures that are consumed from effectful code.
+ */
+class TReentrantLock private (data: TRef[Either[ReadLock, WriteLock]]) {
+
+  /**
+   * Acquires a read lock. The transaction will suspend until no other fibers
+   * are holding write locks. Succeeds with the number of write locks held by
+   * this fiber.
+   */
+  lazy val acquireRead: STM[Nothing, Int] =
+    STM.fiberId.flatMap(fiberId => adjustRead(fiberId, 1))
+
+  /**
+   * Acquires a write lock. The transaction will suspend until no other
+   * fibers are holding read or write locks. Succeeds with the number of
+   * write locks held by this fiber.
+   */
+  lazy val acquireWrite: STM[Nothing, Int] =
+    for {
+      fiberId <- STM.fiberId
+      w <- data.get.collect {
+            case Left(readLock) if (readLock.noOtherHolder(fiberId)) =>
+              WriteLock(1, readLock.readLocks(fiberId), fiberId)
+
+            case Right(WriteLock(n, m, `fiberId`)) =>
```

Comment:
Is there a reason for the backticks?

Reply:
To match for equality for the variable of the same name in the outer scope.